src/lib/operators/join_sort_merge.cpp

```diff
@@ -497,26 +494,47 @@ class JoinSortMerge::JoinSortMergeImpl : public AbstractReadOnlyOperatorImpl {
         _sorted_right_table, [&](RowID right_row_id) { _emit_combination(output_cluster, NULL_ROW_ID, right_row_id); });
   }
 
-  /**
-  * Determines the length of the run starting at start_index in the values vector.
-  * A run is a series of the same value.
-  **/
-  size_t _run_length(size_t start_index, std::shared_ptr<MaterializedSegment<T>> values) {
-    if (start_index >= values->size()) {
+  // Determines the length of the run starting at start_index in the values vector. A run is a series of the same
+  // value. Even though the input vector is sorted, we first start by linearly scanning for the end of the run. The
+  // reason is that we know with a high probability that the item we are looking for is at the beginning of the input
+  // vector. If we do not find the run's end within the linearly scanned part, we use a binary search.
+  size_t _run_length(size_t start_index, const MaterializedSegment<T>& values) {
+    if (start_index >= values.size()) {
       return 0;
     }
 
-    auto start_position = values->begin() + start_index;
-    auto result = std::upper_bound(start_position, values->end(), *start_position,
-                                   [](const auto& a, const auto& b) { return a.value < b.value; });
+    const auto begin = values.begin() + start_index;
+    const auto run_value = begin->value;
+
+    constexpr auto LINEAR_SEARCH_ITEMS = size_t{128};
+    auto end = begin + LINEAR_SEARCH_ITEMS;
+    if (values.end() < end) {
+      // Set end of linear search to end of input vector if we would overshoot otherwise.
+      end = values.end();
+    }
+
+    const auto linear_search_result = std::find_if(begin, end, [&](const auto& v) { return v.value > run_value; });
+    if (linear_search_result != end) {
+      // Match found within the linearly scanned part.
+      return std::distance(begin, linear_search_result);
+    }
+
+    if (linear_search_result == values.end()) {
+      // We found the value neither in the linearly scanned part nor in the input vector. That means that all values
```

Comment:
Am I correct that this means `values.size()` is exactly `LINEAR_SEARCH_ITEMS + start_index`? If yes, I would mention in the comment for clarification that the linear search spanned the whole vector.

Reply:
No, I think it means that the run_length is >= LINEAR_SEARCH_ITEMS. But the values vector can still be much larger than LINEAR_SEARCH_ITEMS.